paddlenlp/trainer/auto_trainer.py

```diff
@@ -0,0 +1,183 @@
+# Copyright (c) 2023 PaddlePaddle Authors. All Rights Reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+from typing import Any, Dict, Optional, Union
+
+import numpy as np
+import paddle
+import paddle.distributed as dist
+import paddle.nn as nn
+from paddle.distributed import fleet
+
+from paddlenlp.trainer import Trainer
+
+from ..utils.log import logger
+from .trainer_utils import _exec_mode_guard, has_length
+
+
+class SemiAutoTrainer(Trainer):
+    def __init__(self, *args, **kwargs):
+
+        if kwargs.get("args", None) is not None and kwargs["args"].run_static_semi_auto:
+            if kwargs.get("criterion", None) is None:
+
+                def loss_func(loss, outputs):
+                    return loss
+
+                kwargs.update({"criterion": loss_func})
+
+        super().__init__(*args, **kwargs)
+        assert self.args.use_auto_parallel
+
+    def _nested_gather(self, tensors):
+        """
+        Gather value of `tensors` (tensor or list/tuple of nested tensors) and convert them to numpy before
+        concatenating them to `gathered`
+        """
+        return tensors
+
+    def _wrap_model(self, model, training=True):
+        self.optimizer = dist.shard_optimizer(self.optimizer) if not self.args.run_static_semi_auto else self.optimizer
+
+        return model
+
+    def _get_meshes_for_loader(self):
+        def _get_mesh(pp_idx=0):
+            mesh = fleet.auto.get_mesh()
+            if "pp" in mesh.dim_names:
+                mesh = mesh.get_mesh_with_dim("pp")[pp_idx]
+            return mesh
+
+        meshes = []
+        for pp_idx in range(self.args.pipeline_parallel_degree):
+            meshes.append(_get_mesh(pp_idx))
+        return meshes
+
+    def _wrap_dist_loader(self, train_dataloader):
+        return dist.shard_dataloader(
+            dataloader=train_dataloader,
+            meshes=self._get_meshes_for_loader(),
+            shard_dims="dp",
+        )
+
+    def _wrap_for_static(self, model, train_dataloader):
```

Comment:
it seems not used?

Reply:
It's called in `Trainer` from`paddlenlp/trainer/trainer.py`, for wrapping `model` into `DistModel` in static mode